### PR TITLE
PXP-5613 Feat/cognito

### PR DESCRIPTION
**Pull request by vpsx**
userdatamodel PR: https://github.com/uc-cdis/userdatamodel/pull/71

### New Features
- Add Cognito as Fence IdP

### Dependency updates
- Bump userdatamodel to 2.3.2

### Deployment changes
- To use the new Cognito integration, apply in fence config the changes made to config-default.yaml. Otherwise no action required.

### CODE REVIEW

File: fence/__init__.py [hunk 1]
```diff
@@ -342,6 +343,14 @@ def _setup_oidc_clients(app):
             logger=logger,
         )
 
+    # Add OIDC client for Amazon Cognito if configured.
+    if "cognito" in oidc:
+        app.cognito_client = CognitoClient(
+            config["OPENID_CONNECT"]["cognito"],
```
Comment (fantix) [inline fence/__init__.py:349]: (trivial)
```python
             oidc["cognito"],
```

File: fence/resources/openid/idp_oauth2.py [hunk 2]
```diff
@@ -84,10 +84,10 @@ def get_value_from_discovery_doc(self, key, default_value):
                     )
                 )
                 return_value = default_value
-            elif return_value != default_value:
+            elif return_value != default_value and default_value != "":
```
Comment (fantix) [inline fence/resources/openid/idp_oauth2.py:87]: (trivial) https://docs.python.org/3/reference/expressions.html#comparisons

File: fence/resources/openid/cognito_oauth2.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        """
+        try:
+            token_endpoint = self.get_value_from_discovery_doc("token_endpoint", "")
+            jwks_endpoint = self.get_value_from_discovery_doc("jwks_uri", "")
+            claims = self.get_jwt_claims_identity(token_endpoint, jwks_endpoint, code)
+
+            if claims["email"] and (
+                claims["email_verified"] or self.settings["assume_emails_verified"]
+            ):
+                return {"email": claims["email"]}
```
Comment (fantix) [inline fence/resources/openid/cognito_oauth2.py:53]: do we want to store all the claims in the database for debugging purposes?

File: fence/__init__.py (hunk 1 above)
Comment (vpsx) [reply fence/__init__.py:349]: 😂 🤦‍♂ 🤦‍♂ 🤦‍♂ Thank you!!

File: fence/resources/openid/idp_oauth2.py (hunk 2 above)
Comment (vpsx) [reply fence/resources/openid/idp_oauth2.py:87]: Chain them to save an eval? 👍

File: fence/resources/openid/cognito_oauth2.py (hunk 3 above)
Comment (vpsx) [reply fence/resources/openid/cognito_oauth2.py:53]: To clarify: The same way as with `post_login` in `SynapseCallback`?

File: fence/resources/openid/idp_oauth2.py (hunk 2 above)
Comment (vpsx) [reply fence/resources/openid/idp_oauth2.py:87]: Note: Per "offline" (slack) discussion leaving as is for readability

File: fence/resources/openid/cognito_oauth2.py (hunk 3 above)
Comment (vpsx) [reply fence/resources/openid/cognito_oauth2.py:60]: Note: per "offline" (slack) discussion using logs instead

**Pull request merged**